lib/Service/TokenService.php

```diff
@@ -0,0 +1,276 @@
+use OCP\IURLGenerator;
+use OCP\IUserSession;
+use OCP\PreConditionNotMetException;
+use OCP\Security\ICrypto;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Token management service
+ * This is helpful to debug:
+ * tail -f data/nextcloud.log | grep "\[Token" | jq ".time,.message"
+ */
+class TokenService {
+
+	private const SESSION_TOKEN_KEY = Application::APP_ID . '-user-token';
+
+	private IClient $client;
+
+	public function __construct(
+		IClientService $clientService,
+		private ISession $session,
+		private IUserSession $userSession,
+		private IConfig $config,
+		private LoggerInterface $logger,
+		private ICrypto $crypto,
+		private IRequest $request,
+		private IURLGenerator $urlGenerator,
+		private DiscoveryService $discoveryService,
+		private ProviderMapper $providerMapper,
+	) {
+		$this->client = $clientService->newClient();
+	}
+
+	public function storeToken(array $tokenData): Token {
+		$token = new Token($tokenData);
+		$this->session->set(self::SESSION_TOKEN_KEY, json_encode($token, JSON_THROW_ON_ERROR));
+		$this->logger->debug('[TokenService] Store token');
+		return $token;
+	}
+
+	/**
+	 * Get the token stored in the session
+	 * If it has expired: try to refresh it
+	 *
+	 * @param bool $refreshIfExpired
+	 * @return Token|null Return a token only if it is valid or has been successfully refreshed
+	 * @throws \JsonException
+	 */
+	public function getToken(bool $refreshIfExpired = true): ?Token {
+		$sessionData = $this->session->get(self::SESSION_TOKEN_KEY);
+		if (!$sessionData) {
+			$this->logger->debug('[TokenService] getToken: no session data');
+			return null;
+		}
+
+		$token = new Token(json_decode($sessionData, true, 512, JSON_THROW_ON_ERROR));
+		// token is still valid
+		if (!$token->isExpired()) {
+			$this->logger->debug('[TokenService] getToken: token is still valid, it expires in ' . $token->getExpiresInFromNow() . ' and refresh expires in ' . $token->getRefreshExpiresInFromNow());
+			return $token;
+		}
+
+		// token has expired
+		// try to refresh the token if the refresh token is still valid
+		if ($refreshIfExpired && !$token->refreshIsExpired()) {
+			$this->logger->debug('[TokenService] getToken: token is expired and refresh token is still valid, refresh expires in ' . $token->getRefreshExpiresInFromNow());
+			return $this->refresh($token);
+		}
+
+		$this->logger->debug('[TokenService] getToken: return a token that has not been refreshed');
+		return $token;
+	}
+
+	/**
+	 * Check to make sure the login token is still valid
+	 *
+	 * @return void
+	 * @throws \JsonException
+	 * @throws PreConditionNotMetException
+	 */
+	public function checkLoginToken(): void {
+		$currentUser = $this->userSession->getUser();
+		if (!$this->userSession->isLoggedIn() || $currentUser === null) {
+			$this->logger->debug('[TokenService] checkLoginToken: user not logged in');
+			return;
+		}
+		if ($this->config->getUserValue($currentUser->getUID(), Application::APP_ID, 'had_token_once', '0') !== '1') {
```

Comment:
I think I miss understanding on why we need this extra user value check, maybe you can elaborate on that?

Reply:
If the user never had any login token before, it means it's not using user_oidc to authenticate so we know we can skip the login token check.

Do you mean that it is enough to just do nothing if there is no login token in the session?
We need to know if there is no login token available
* because user_oidc was not used to log in (we can silently skip the check)
* OR because the session has died/disappeared (then we need to reauthenticate)

Does that make sense?

Reply:
Ah ok, makes sense then 👍